assets/js/base.js

```diff
@@ -209,5 +217,127 @@ window.Herman = (function base(Herman, $) {
+    let matches = $();
+    if (results && results.length) {
+      for (const res of results) {
+        const doc = Herman.searchStore[res.ref];
+        const toMark = {
+          title: [],
+          contents: [],
+        };
+        Object.keys(res.matchData.metadata).forEach(term => {
+          Object.keys(res.matchData.metadata[term]).forEach(fieldName => {
+            // For each term found in each field, store position of matches
+            const pos = res.matchData.metadata[term][fieldName].position.map(
+              p => ({
+                start: p[0],
+                length: p[1],
+              })
+            );
+            // @@@ Temporary solution until fieldname bug is fixed:
+            // https://github.com/olivernn/lunr.js/issues/320
+            const field = fieldName === 'title' ? 'title' : 'contents';
+            toMark[field] = toMark[field].concat(pos);
+          });
+        });
+        if (!(toMark.title.length || toMark.contents.length)) {
+          return;
+        }
+        const obj = {
+          url: `/${res.ref}`,
+          title: doc.title,
+          contents: toMark.contents.length ? doc.contents : '',
+        };
+        // Render search result template
+        const el = $(window.nunjucks.render('search_result.j2', obj));
+        if (toMark.title.length) {
+          // Highlight matches in `title` field
+          el.find(`[data-result-field="title"]`).markRanges(toMark.title);
+        }
+        if (toMark.contents.length) {
+          const textEl = el.find(`[data-result-field="contents"]`);
+          toMark.contents.sort((a, b) => a.start - b.start);
+          // Highlight first 5 matches in `contents` field
+          textEl.markRanges(toMark.contents.slice(0, 5), {
+            done: () => {
+              // Truncate text not within 15 words of a match
+              textEl.get(0).childNodes.forEach(node => {
+                const hasPrev = node.previousSibling !== null;
+                const hasNext = node.nextSibling !== null;
+                const isText = node.nodeName === '#text';
+                if (isText) {
+                  const text = node.nodeValue.split(' ');
+                  if (hasPrev && hasNext) {
+                    if (text.length > 30) {
+                      text.splice(15, text.length - 30, '…');
+                    }
+                  } else if (hasNext) {
+                    if (text.length > 15) {
+                      text.splice(0, text.length - 15, '…');
+                    }
+                  } else if (hasPrev) {
+                    if (text.length > 15) {
+                      text.splice(15, text.length - 15, '…');
+                    }
+                  }
+                  node.nodeValue = text.join(' ');
+                }
+              });
+            },
+          });
+        }
+        // Add search result template to set of results
+        matches = matches.add(el);
+      }
```

Comment:
It's not urgent, but if `showResults` could be decomposed into well-named things, that'd probably help. It's a bit hard to follow.